test/Chainweb/Test/Pact/RemotePactTest.hs

```diff
@@ -288,94 +284,68 @@ withRequestKeys iot ioNonce networkIO f = withResource mkKeys (\_ -> return ())
-            Right _ -> do
-                debug $ "send succeeded after " ++ show (maxSendRetries - retries) ++ " retries"
-                return result
-
-maxPollRetries :: Int
-maxPollRetries = 30
-
--- | To allow time for node to startup, retry a number of times
-pollWithRetry
+    ss = _sbCmds batch
+
+-- | Poll with retry using an exponential backoff
+--
+polling
     :: PactTestApiCmds
     -> ClientEnv
     -> RequestKeys
     -> IO PollResponses
-pollWithRetry cmds env rks = do
-  sleep 3
-  go maxPollRetries
-    where
-      go !retries = do
-          let retry failure = do
-                if retries == 0
-                  then do
-                    debug $ "poll failing after " ++ show maxSendRetries
-                               ++ " retries"
-                    fail failure
-                  else sleep 1 >> go (retries - 1)
-          let rkks = _rkRequestKeys rks
-          result <- runClientM (pollApiCmd cmds (Poll rkks)) env
-          case result of
-              Left e -> retry (show e)
-              Right (PollResponses mp) -> do
-                  ok <- checkResults (toList rkks) mp
-                  if ok
-                    then do
-                        debug $ concat
-                            [ "poll succeeded after "
-                            , show (maxSendRetries - retries)
-                            , " retries" ]
-                        return $ PollResponses mp
-                    else do
-                        debug "poll check failed, continuing"
-                        let msg = concat [ "poll check failure. keys: "
-                                         , show rks
-                                         , "\nresult map was: "
-                                         , show mp
-                                         ]
-                        retry msg
-
-      checkCommandResult x cr =
-          _crReqKey cr == x && isOkPactResult (_crResult cr)
-
-      isOkPactResult (PactResult e) = isRight e
-
-      checkResults [] _ = return True
-      checkResults (x:xs) mp = case HashMap.lookup x mp of
-                                 (Just cr) -> if checkCommandResult x cr
-                                                then checkResults xs mp
-                                                else return False
-                                 Nothing -> return False
+polling api cenv rks =
+    recoverAll (exponentialBackoff 10000 <> limitRetries 15) $ \s -> do
+      putStrLn
+        $ "polling for requestkeys " <> show (toList rs)
+        <> " [" <> show (view rsIterNumberL s) <> "]"
+
+      -- Run the poll cmd loop and check responses
+      -- by making sure results are successful and request keys
+      -- are sane
+
+      runClientM (pollApiCmd api $ Poll rs) cenv >>= \case
+        Left e -> fail $ show e
+        Right r@(PollResponses mp) ->
+          if all (go mp) (toList rs)
+          then return r
+          else fail "polling check failed"
```

Comment:
Is there a cooler thing to do than `fail`?

Reply:
not really. I just need to signal an error at this point to trigger a retry

Reply:
`fail` is fine here